vivid-client/src/components/Canvas.tsx

```diff
@@ -1,82 +1,82 @@
-      client.emit('ready');
-      clientState;
-    });
+  function init() {
+    if (!canvasRef.current) return;
 
-    client.on('init', () => {
-      if (!canvasRef.current) return;
+    canvas = canvasRef.current;
+    context = canvas.getContext('2d');
 
-      canvas = canvasRef.current;
-      context = canvas.getContext('2d');
+    document.addEventListener('keydown', keydown);
+    document.addEventListener('keyup', keyup);
+    document.addEventListener('mousemove', mouseMove);
+  }
 
-      document.addEventListener('keydown', keydown);
-      document.addEventListener('keyup', keyup);
-      document.addEventListener('mousemove', mouseMove);
+  function draw(gameState: IGameState) {
+    if (context === null) return;
+    requestAnimationFrame(() => {
+      drawGame(gameState, canvas, context);
     });
+  }
 
-    client.on('drawGame', (gameState: IGameState) => {
-      if (context === null) return;
-      requestAnimationFrame(() => {
-        drawGame(gameState, canvas, context);
-      });
-    });
+  function gameOver(winner: string) {
+    alert(`User: "${winner}" won the game`);
+  }
 
-    client.on('disconnect', () => {
-      setClientState('DISCONNECTED');
-    });
+  function start(roomName: string) {
+    client.emit('start', roomName);
+  }
 
-    client.on('gameOver', (winner: string) => {
-      alert(`User: "${winner}" won the game`);
-    });
-
-    client.on('start', (roomName: string) => {
-      client.emit('start', roomName);
-    });
+  function keydown(event: KeyboardEvent) {
+    if (event.key === 'w') client.emit('keydown', -0.01);
+    else if (event.key === 's') client.emit('keydown', 0.01);
+  }
 
-    function keydown(event: KeyboardEvent) {
-      if (event.key === 'w') client.emit('keydown', -0.01);
-      else if (event.key === 's') client.emit('keydown', 0.01);
-    }
+  function keyup(event: KeyboardEvent) {
+    if (event.key === 'w' || event.key === 's') client.emit('keydown', 0);
+  }
 
-    function keyup(event: KeyboardEvent) {
-      if (event.key === 'w' || event.key === 's') client.emit('keydown', 0);
-    }
+  function mouseMove(event: MouseEvent) {
+    client.emit('mouseMove', event.clientY / canvas.height);
+  }
 
-    function mouseMove(event: MouseEvent) {
-      client.emit('mouseMove', event.clientY / canvas.height);
-    }
+  React.useEffect(() => {
+    if (!client) return;
+    client.on('connect', connect);
+    client.on('init', init);
+    client.on('start', start);
+    client.on('drawGame', draw);
+    client.on('gameOver', gameOver);
 
     return () => {
-      client.destroy();
+      client.off('connect', connect);
+      client.off('init', init);
+      client.off('start', start);
+      client.off('drawGame', draw);
+      client.off('gameOver', gameOver);
       document.removeEventListener('keydown', keydown);
       document.removeEventListener('keyup', keyup);
       document.removeEventListener('mousemove', mouseMove);
     };
-  }, []);
+  }, [client]);
```

Comment:
dont return the effect prematurely. otherwise there is a change it doesnt remove the event listeners.
so do `if (client) {}` on both the client.off's and the client.on's